Add Traverse Docs (#1534)
arrow-kt/arrow: modules/docs/arrow-docs/docs/docs/arrow/typeclasses/traverse/README.md

{:.intermediate}
intermediate

> In functional programming it is very common to encode "effects" as data types - common effects include `Option` for possibly missing values, `Either` and `Validated` for possible errors, and [`Promise`]({{ '/docs/effects/promise/' | relative_url }}) for asynchronous computations, which has the same purpose as a `Future`, only pure and lazy.

Reviewer: "effects" is a word we use for other meaning. Use "behaviors" instead.

Reviewer: Not resolved IMO.

Given just `(User) -> Promise<ForIO, Profile>`, what should we do if we want to fetch profiles for a `List<User>`? We could try familiar combinators like map.

```kotlin:ank
fun profilesFor(users: List<User>): List<Promise<ForIO, Profile>> = users.map(::userInfo)
```

Note the return type `List<Promise<ForIO, Profile>>`. This makes sense given the type signatures, but seems unwieldy. We now have a list of asynchronous values, and to work with those values we must then use the combinators on `Promise` for every single one. It would be nicer instead if we could get the aggregate result in a single `Promise`, say a `Promise<ForIO, List<Profile>>`.

There exists a much more generalized form that would allow us to parse a `List<String>` or validate credentials for a `List<User>`.

Enter `Traverse`.

### The Typeclass

> At center stage of Traverse is the traverse method.

Reviewer: `traverse` method. Between backticks

```kotlin
import arrow.typeclasses.Applicative
import arrow.typeclasses.Foldable
import arrow.typeclasses.Functor

interface Traverse<F> : Functor<F>, Foldable<F> {
  fun <G, A, B> Kind<F, A>.traverse(AP: Applicative<G>, f: (A) -> Kind<G, B>): 
}
```

> **Essentially, one uses `Traverse`, when there is a function `(A) -> Kind<G, B>` and `G` is an `Applicative` instance e.g.: `(User) -> Promise<ForIO, Profile>` and you want to `lift` the previous function to work on `Kind<F, A>` where `F` is a `Traverse` instance e.g.: `List<User>` and return `Kind<G, Kind<F, B>>` e.g.: `Promise<ForIO, List<Profile>>`**

Reviewer: This paragraph is too dense. I would split it.

Author: I deleted this section because the next paragraph explains the same concept, just with more concrete Datatypes. I am shifting away from this style of writing.

In our above example, `F` is `List`, and `G` is `Option`, `Either`, or `Promise`. For the profile example, traverse says given a `List<User>` and a function `(User) -> Promise<ForIO, Profile>`, it can give you a `Promise<ForIO, List<Profile>>`.

Abstracting away the `G` (still imagining `F` to be `List`), `traverse` says given a collection of data, and a function that takes a piece of data and returns an effectful value, it will traverse the collection, applying the function and aggregating the effectful values (in a `List`) as it goes.

In the most general form, `Kind<F, A>` is some sort of context `F` which may contain a value `A` (or several). While `List` tends to be among the most general cases, there also exist `Traverse` instances for `Option`, `Either`, and `Validated` (among others).

#### Why not Foldable

> 

Reviewer: Empty section?

> #### Choose your effect

Author: @pakoito Should it be 
```
### Choose your data type
```
Or something else

Reviewer: ### Choose your implementation ?

Author: I don't have a better idea, I would go for your suggestion :)

> The type signature of `Traverse` appears highly abstract, and indeed it is - what `traverse` does as it walks the `Kind<F, A>` depends on the effect of the function. Let's see some examples where `F` is taken to be `List`.

Reviewer: `, and indeed it is` --> `, although it's easier if you think about collections of executable operations`

```kotlin:ank
import arrow.core.ValidatedNel
import arrow.core.invalidNel
import arrow.core.toT
import arrow.core.validNel

> import arrow.core.Either

Reviewer: This one is carried from a prior snippet.

fun main() {
  //sampleStart
  val validatedList = listOf("1", "22w", "3", "33s").k()
    .traverse(ValidatedNel.applicative(Nel.semigroup<NumberFormatException>()), ::parseIntValidated).fix()
  //sampleEnd
  println(validatedList)
}
```

Notice that in the `Either` case, should any string fail to parse the entire `traversal` is considered a failure. Moreover, once it hits its first bad parse, it will not attempt to parse any others down the line (similar behavior would be found with using `Option` as the effect). Contrast this with `Validated` where even if one bad parse is hit, it will continue trying to parse the others, accumulating any and all errors as it goes. The behavior of traversal is closely tied with the `Applicative` behavior of the data type.

Going back to our `Promise` example, we can get an `Applicative` instance for `IO`, by converting `Promise` to `IO` that runs each `Promise` concurrently. Then when we traverse a `List<A>` with an `(A) -> Promise<ForIO, B>`, we can imagine the traversal as a scatter-gather. Each `A` creates a concurrent computation that will produce a `B` (the scatter), and as the `Promise`s complete they will be gathered back into a `List`.

> {TODO: finish example with Promise}

Reviewer: TODO

```kotlin:ank:playground
import arrow.core.Either
import arrow.core.Option
import arrow.fx.Promise

interface Profile
interface User

fun parseInt(s: String): Option<Int> = TODO()
data class SecurityError(val message: String)
data class Credentials(val email: String, val password: String)

fun validateLogin(c: Credentials): Either<SecurityError, Unit> = TODO()

fun userInfo(u: User): Promise<ForIO, Profile> = TODO()

data class Admin(val name: String, val crd: Credentials): User

fun ListK<User>.processLogin()
```

> Evidently, `Traverse` is not limited to `List` or `Nel`, it provides an abstraction over 'things that can be traversed over', like a Binary tree, hence the name `Traverse`.

Reviewer: `like a Binary tree, a Sequence, or a Stream`

#### Playing with `Reader`

> Another interesting effect we can use is Reader. Recall that a `Reader<D, A>` is a type alias for `Kleisli<ForId, D, A>` which is a wrapper around `(D) -> A`.

Reviewer: A bit more context. Reader is an `injector for a parameter` or something similar.

If we fix `D` to be some sort of dependency or configuration, we can use the `Reader` applicative in our `traverse`.

```kotlin:ank
import arrow.mtl.Reader

interface Context
interface Topic
interface Result

typealias Job<A> = Reader<Context, A>

fun processTopic(t: Topic): Job<Result> = TODO()
```

We can imagine we have a data pipeline that processes a bunch of data, each piece of data being categorized by a topic. Given a specific topic, we produce a `Job` that processes that topic. (Note that since a `Job` is just a `Reader`/`Kleisli`, one could write many small `Jobs` and compose them together into one `Job` that is used/returned by `processTopic`.)

Corresponding to our bunches of data are bunches of topics, a `List<Topic>` if you will. Since Reader has an `Applicative` instance, we can `traverse` over this list with `processTopic`.

```kotlin:ank
import arrow.core.Id
import arrow.core.ForId
import arrow.core.ListK
import arrow.core.extensions.id.applicative.applicative
import arrow.mtl.KleisliPartialOf
import arrow.mtl.extensions.kleisli.applicative.applicative
import arrow.mtl.fix

fun processTopics(topics: ListK<Topic>): Job<ListK<Result>> =

>   topics.traverse<KleisliPartialOf<ForId, Context>, Result>(Job.applicative<ForId, Context>(Id.applicative())) {

Reviewer: I'd extract the instance `Job.applicative<ForId, Context>(Id.applicative())` to a field here.

    processTopic(it)
  }.fix()
```

Note the nice return type - `Job<List<Result>>`. We now have one aggregate `Job` that when run, will go through each topic and run the topic-specific job, collecting results as it goes. We say "when run" because a `Job` is some function that requires a Context before producing the value we want.

> One example of a "context" can be found in the [Spark](http://spark.apache.org/) project. In Spark, information needed to run a Spark job (where the master node is, memory allocated, etc.) resides in a `SparkContext`. Going back to the above example, we can see how one may define topic-specific Spark jobs `(type Job<A> = Reader<SparkContext, A>)` and then run several Spark jobs on a collection of topics via `traverse`. We then get back a `Job<List<Result>>`, which is equivalent to (`SparkContext) -> List<Result>`. When finally passed a `SparkContext`, we can run the job and get our results back.

Reviewer: I'd also add an example of Android's Context.

Author: I purpose to add a more tagless final approach, so Users are not inclined to use MTL.
What do you think @pakoito ?

Moreover, the fact that our aggregate job is not tied to any specific `SparkContext` allows us to pass in a `SparkContext` pointing to a production cluster, or (using the exact same job) pass in a test `SparkContext` that just runs locally across threads. This makes testing our large job nice and easy.

Finally, this encoding ensures that all the jobs for each topic run on the exact same cluster. At no point do we manually pass in or thread a `SparkContext` through - that is taken care for us by the (applicative) effect of `Reader` and therefore by `traverse`.

> ### Sequencing

Reviewer: TBH given the number of requests we get I'd start with sequencing, and then explaining that traverse is a specialization.

Reviewer 2: Exactly. The most usual problem is the one about "reversing the types" which is what sequence resolves. You can start with sequence, then in the end unveil its equivalent to traverse with identity, then explain Traverse and how it's even more powerful being able to use other transformation functions.

import arrow.core.extensions.option.applicative.map
import arrow.core.fix
import arrow.core.identity
import arrow.core.none
import arrow.core.some
import arrow.core.Option

fun main() {
  //sampleStart
  val optionList: Option<List<Int>> =
    listOf(1.some(), 2.some(), 3.some())
      .traverse(Option.applicative(), ::identity).fix().map { it.fix() }

  val emptyList: Option<List<Int>> =
    listOf(1.some(), none(), 3.some())
      .traverse(Option.applicative(), ::identity).map { it.fix() }
  //sampleEnd
  println("optionList = $optionList")
  println("emptyList = $emptyList")
}
```

`Traverse` provides a convenience method `sequence` that does exactly this.

```kotlin:ank:playground
import arrow.core.Option
import arrow.core.extensions.list.traverse.sequence
import arrow.core.extensions.option.applicative.applicative
import arrow.core.none
import arrow.core.some

fun main() {
  //sampleStart
  val optionList =
    listOf(1.some(), 2.some(), 3.some())
      .sequence(Option.applicative())

  val emptyList =
    listOf(1.some(), none(), 3.some())
      .sequence(Option.applicative())
  //sampleEnd
  println("optionList = $optionList")
  println("emptyList = $emptyList")
}
```

> -- Add Text

Reviewer: TODO

### Traversables are Functors

> {TODO: finish this section}

Reviewer: TODO

> ### Traversing for effect

Reviewer: sequence_ is in foldable, isn't it?

Author: yes

```

We end up with a `IO<ListK<Unit>>`! A `ListK<Unit>` is not of any use to us, and communicates the same amount of information as a single `Unit` does.

Traversing solely for the sake of the effect (ignoring any values that may be produced, `Unit` or otherwise) is common, so `Foldable` (superclass of `Traverse`) provides `traverse_` and `sequence_` methods that do the same thing as `traverse` and `sequence` but ignores any value produced along the way, returning `Unit` at the end.

```kotlin:ank
fun writeManyToStore(data: ListK<Data>): IO<Unit> =
  data.traverse_(IO.applicative()) { writeToStore(it).get() }.fix()
```
```kotlin:ank:playground
import arrow.core.Option
import arrow.core.extensions.list.foldable.sequence_
import arrow.core.extensions.option.applicative.applicative
import arrow.core.none
import arrow.core.some

fun main() {
  //sampleStart
  val someUnit = listOf(1.some(), 2.some(), 3.some())
    .sequence_(Option.applicative())
  val noneUnit = listOf(1.some(), none(), 3.some())
    .sequence_(Option.applicative())
  //sampleEnd
  println("someUnit= $someUnit")
  println("noneUnit= $noneUnit")
}
```

### Traversables are Foldable

> The `Foldable` type class abstracts over “things that can be folded over” similar to how `Traverse` abstracts over “things that can be traversed.” It turns out `Traverse` is strictly more powerful than `Foldable` - that is, `foldLeft` and `foldRight` can be implemented in terms of `traverse` by picking the right `Applicative`. However, arrow's `Traverse` does not implement `foldLeft` and `foldRight` as the actual implementation tends to be ineffecient, as we one can implement the `Foldable` `fold`'s.

Author: This is just a guess, but I am guessing that the Foldable fold is more efficient, because traverse can blowUp the stack if it isn't defined by fold. 
https://github.com/typelevel/cats/issues/1473

For brevity and demonstration purposes we’ll implement an isomorphic `foldMap` method in terms of `traverse` by using `arrow.typeclasses.Const`. You can then implement `foldRight` in terms of `foldMap`, and `foldLeft` can then be implemented in terms of `foldRight`, though the resulting implementations may be slow.

```kotlin:ank

> // TODO: finish this section

Reviewer: TODO

{:.intermediate}
intermediate

In functional programming it is very common to encode "behaviors" as data types - common behaviors include `Option` for possibly missing values, `Either` and `Validated` for possible errors, and [`Promise`]({{ '/docs/effects/promise/' | relative_url }}) for asynchronous computations, which has the same purpose as a `Future`, only pure and lazy.

These behaviors tend to show up in functions working on a single piece of data - for instance parsing a single `String` into an `Int`, validating a login, or asynchronously fetching website information for a user.

```kotlin:ank
import arrow.core.Either
import arrow.core.Option
import arrow.fx.ForIO
import arrow.fx.Promise

interface Profile
interface User

> fun userInfo(u: User): Promise<ForIO, Profile> = TODO()

Reviewer 3: This is just some nitpicking but `userInfo` would be better suitable for a field name. What do you think about `loadUserInfo`, or something similar?

```

> Each function asks only for the data it actually needs; in the case of `userInfo`, a single `User`. We certainly could write one that takes a `List<User>` and fetch profile for all of them, would be a bit strange. If we just wanted to fetch the profile of just one user, we would either have to wrap it in a `List` or write a separate function that takes in a single user anyways. More fundamentally, functional programming is about building lots of small, independent pieces and composing them to make larger and larger pieces - does this hold true in this case?

Reviewer 3: ```suggestion
Each function asks only for the data it actually needs; in the case of `userInfo`, a single `User`. We could certainly write one which takes a `List<User>` and fetches a profile for all of them, but it would be a bit strange. If we wanted to fetch the profile of just one user, we would either have to wrap it in a `List` or write a separate function that takes in a single user anyways. More fundamentally, functional programming is about building lots of small, independent pieces and composing them to make larger and larger pieces - does this hold true in this case?
```

Author: thx

Given just `(User) -> Promise<ForIO, Profile>`, what should we do if we want to fetch profiles for a `List<User>`? We could try familiar combinators like map.

```kotlin:ank
fun profilesFor(users: List<User>): List<Promise<ForIO, Profile>> = users.map(::userInfo)
```

Note the return type `List<Promise<ForIO, Profile>>`. This makes sense given the type signatures, but seems unwieldy. We now have a list of asynchronous values, and to work with those values we must then use the combinators on `Promise` for every single one. It would be nicer instead if we could get the aggregate result in a single `Promise`, say a `Promise<ForIO, List<Profile>>`.

There exists a much more generalized form that would allow us to parse a `List<String>` or validate credentials for a `List<User>`.

Enter `Traverse`.

### The Typeclass

At center stage of Traverse is the `traverse` method.

```kotlin
import arrow.Kind
import arrow.typeclasses.Applicative
import arrow.typeclasses.Foldable
import arrow.typeclasses.Functor

interface Traverse<F> : Functor<F>, Foldable<F> {
  fun <G, A, B> Kind<F, A>.traverse(AP: Applicative<G>, f: (A) -> Kind<G, B>): Kind<G, Kind<F, B>>
}
```

> In our above example, `F` is `List`, and `G` is `Option`, `Either`, or `Promise`. For the profile example, traverse says given a `List<User>` and a function `(User) -> Promise<ForIO, Profile>`, it can give you a `Promise<ForIO, List<Profile>>`.

Reviewer 4: So `F` could be named a `container`, am I right?

Reviewer 3: I'd love to have readable type parameters instead of `G`, `A`, `B` which convey no information in themselves.

Author: They have to be polymorphic :) @adam-arold
Yes @rcd27 , `F` can be seen as a "container", to be precise it is a parameter for type constructors, e.g.: containers like List, Either..

Reviewer 3: @i-walker what I meant is that you could use `CONTAINER` instead of `T` in the example above. The point is that the generic type parameter would convey what it stands for not just an arbitrary letter like `F`.

### To fold Or not to fold

Even though, `Foldable` and `Traverse` are related, because both 'reduce their values to something', it is not obvious why to consider `Traverse` over `Foldable`.

Here is a small example:

```kotlin:ank:playground
import arrow.core.MapK
import arrow.core.Option
import arrow.core.Some
import arrow.core.extensions.option.applicative.applicative
import arrow.core.fix
import arrow.core.k

fun main(){
  //sampleStart
  val map: MapK<String, Int> = mapOf("one" to 1, "two" to 2).k()

  val optionMapK: Option<MapK<String, String>> = map.traverse(Option.applicative()) { Some("$it") }.fix()
  
  // val optionMapKBoilered = map.foldLeft(Some(emptyMap())) { acc: Option<MapK<String, String>>, i: Int ->
  //   acc.fold({ emptyMap() }, { /*Some logic to retrieve the key of a value, transform it and add it to the accumulated Map*/ })
  // }
  //sampleEnd
  println(optionMapK)
}
```

> Both values try to attain the same thing, but what `Foldable` lacks is that it solely drills down to it's `A` here `Int` and does not preserve it's shape `F` - `MapK`. This is where `Traverse` shines, whenever you care about the Output `B` from `(A) -> B` and the existing shape of `F` you would use `traverse`.

Reviewer 3: This is a great comparison! These nuggets of information are the ones which I'm missing from the rest of the docs! Great work! 👍

> Additionally, your able to wrap your context `F` within a `G`. That is, may among others, the reason why `Traverse` is strictly more powerful than `Foldable`.

Reviewer 3: ```suggestion
Additionally, your are able to wrap your context `F` within a `G`. That is, among others, the reason why `Traverse` is strictly more powerful than `Foldable`.
```

> Additionally, your able to wrap your context `F` within a `G`. That is, may among others, the reason why `Traverse` is strictly more powerful than `Foldable`.

Reviewer 3: I'd add some counterexamples to justify the existence of `Foldable`. If `Traverse` is a strict upgrade why would I use `Foldable`?

Author: there is a section down below which explains that

  val isLeft = listOf("1", "abc", "3").k().traverse(Either.applicative(), ::parseIntEither).isEmpty()
  //sampleEnd
  println("list= $list")
  println("isLeft= $isLeft")
}
```

```kotlin:ank:playground
import arrow.core.Either
import arrow.core.Nel
import arrow.core.ValidatedNel
import arrow.core.extensions.nonemptylist.semigroup.semigroup
import arrow.core.extensions.validated.applicative.applicative
import arrow.core.fix
import arrow.core.invalidNel
import arrow.core.k
import arrow.core.validNel

fun parseIntEither(s: String): Either<NumberFormatException, Int> =
  if (s.matches(Regex("-?[0-9]+"))) Either.Right(s.toInt())
  else Either.Left(NumberFormatException("$s is not a valid integer."))

fun parseIntValidated(s: String): ValidatedNel<NumberFormatException, Int> =
  if (s.matches(Regex("-?[0-9]+"))) s.toInt().validNel()
  else NumberFormatException("$s is not a valid integer.").invalidNel()

fun main() {
  //sampleStart
  val validatedList = listOf("1", "22w", "3", "33s").k()
    .traverse(ValidatedNel.applicative(Nel.semigroup<NumberFormatException>()), ::parseIntValidated).fix()
  //sampleEnd
  println(validatedList)
}
```

> Notice that in the `Either` case, should any string fail to parse the entire `traversal` is considered a failure. Moreover, once it hits its first bad parse, it will not attempt to parse any others down the line (similar behavior would be found with using `Option`). Contrast this with `Validated` where even if one bad parse is hit, it will continue trying to parse the others, accumulating any and all errors as it goes. The behavior of traversal is closely tied with the `Applicative` behavior of the data type.

Reviewer 3: ```suggestion
Notice that in the `Either` case, should any string fail to parse the entire `traverse` is considered a failure. Moreover, once it hits its first bad parse, it will not attempt to parse any others down the line (similar behavior would be found with using `Option`). Contrast this with `Validated` where even if one bad parse is hit, it will continue trying to parse the others, accumulating any and all errors as it goes. The behavior of traversal is closely tied with the `Applicative` behavior of the data type.
```

> Going back to our `Promise` example, we can get an `Applicative` instance for `IO`, by converting `Promise` to `IO` that runs each `Promise` concurrently. Then when we traverse a `List<A>` with an `(A) -> Promise<ForIO, B>`, we can imagine the traversal as a scatter-gather. Each `A` creates a concurrent computation that will produce a `B` (the scatter), and as the `Promise`s complete they will be gathered back into a `List`.

Reviewer 3: ```suggestion
Going back to our `Promise` example, we can get an `Applicative` instance for `IO`, by converting `Promise` to `IO` that runs each `Promise` concurrently. Then when we traverse a `List<A>` with an `(A) -> Promise<ForIO, B>`, we can imagine the traversal as a divide & conquer. Each `A` creates a concurrent computation that will produce a `B` (divide), and as the `Promise`s complete they will be gathered back into a `List` (conquer).
```

```kotlin:ank:playground
import arrow.core.ListK
import arrow.core.k
import arrow.fx.ForIO
import arrow.fx.IO
import arrow.fx.Promise
import arrow.fx.extensions.io.applicative.applicative
import arrow.fx.extensions.io.applicativeError.attempt
import arrow.fx.extensions.io.async.async
import arrow.fx.extensions.io.monad.flatMap
import arrow.fx.fix

interface Profile
interface User

data class DummyUser(val name: String) : User
data class DummyProfile(val u: User) : Profile

fun main() {
  //sampleStart
  val promise: IO<Promise<ForIO, Profile>> =
    Promise.uncancelable<ForIO, Profile>(IO.async()).fix()

  fun userInfo(u: User): IO<Profile> =
    promise.flatMap { p -> p.complete(DummyProfile(u)).flatMap { p.get() } }

  fun ListK<User>.processLogin() =
    traverse(IO.applicative()) { userInfo(it) }

>   val res = listOf(DummyUser("Micheal"), DummyUser("Juan"), DummyUser("T'Challa")).k()

Author: As a todo I will turn this into an Fx style

{:.intermediate}
intermediate

> In functional programming it is very common to encode "behaviors" as data types - common behaviors include `Option` for possibly missing values, `Either` and `Validated` for possible errors, and [`Promise`]({{ '/docs/effects/promise/' | relative_url }}) for asynchronous computations, which has the same purpose as a `Future`, only pure and lazy.

Reviewer 2: I think we should always add links to all data types / typeclasses we mention first time in an article. People doesn't read docs in order so they'll get to this page sometimes even knowing about other essential stuff. (Meaning to add links for Option, Either, Validated here, not just for Promise)

> In functional programming it is very common to encode "behaviors" as data types - common behaviors include `Option` for possibly missing values, `Either` and `Validated` for possible errors, and [`Promise`]({{ '/docs/effects/promise/' | relative_url }}) for asynchronous computations, which has the same purpose as a `Future`, only pure and lazy.

Author: Done

These behaviors tend to show up in functions working on a single piece of data - for instance parsing a single `String` into an `Int`, validating a login, or asynchronously fetching website information for a user.

```kotlin:ank
import arrow.core.Either
import arrow.core.Option
import arrow.fx.ForIO
import arrow.fx.Promise

interface Profile
interface User

fun userInfo(u: User): Promise<ForIO, Profile> = TODO()
```

> Each function asks only for the data it actually needs; in the case of `userInfo`, a single `User`. Certainly, we could write one that takes a `List<User>` and fetches profile for all of them,but it would be a bit strange. If we just wanted to fetch the profile of just one user, we would either have to wrap it in a `List` or write a separate function that takes in a single user anyways. More fundamentally, functional programming is about building lots of small, independent pieces and composing them to make larger and larger pieces - does this hold true in this case?

Author: Makes sense. I edited it

Enter `Traverse`.

### The Typeclass

At center stage of Traverse is the `traverse` method.

```kotlin
import arrow.Kind
import arrow.typeclasses.Applicative
import arrow.typeclasses.Foldable
import arrow.typeclasses.Functor

interface Traverse<F> : Functor<F>, Foldable<F> {
  fun <G, A, B> Kind<F, A>.traverse(AP: Applicative<G>, f: (A) -> Kind<G, B>): Kind<G, Kind<F, B>>
}
```

In our above example, `F` is `List`, and `G` is `Option`, `Either`, or `Promise`. For the profile example, traverse says given a `List<User>` and a function `(User) -> Promise<ForIO, Profile>`, it can give you a `Promise<ForIO, List<Profile>>`.

Abstracting away the `G` (still imagining `F` to be `List`), `traverse` says given a collection of data, and a function that takes a piece of data and returns an value `B` wrapped in a container `G`, it will traverse the collection, applying the function and aggregating the values (in a `List`) as it goes.

In the most general form, `Kind<F, A>` is some sort of context `F` which may contain a value `A` (or several). While `List` tends to be among the most general cases, there also exist `Traverse` instances for `Option`, `Either`, and `Validated` (among others).

> Essentially, one uses `Traverse`, when there is a function `(A) -> Kind<G, B>` and `G` is an `Applicative` instance, thus `G` provides `just` and `map` functions e.g.: just as `Promise<ForIO, B>` provides and you want to `lift` the previous function to work on `Kind<F, A>` where `F` is a `Traverse` instance - and you still care about `A` (the `Profiles`) `(List<User>)` and return `Kind<G, Kind<F, B>>` e.g.: `Promise<ForIO, List<Profile>>`

Author: I choose to remove it. It really creates more overhead then it helps. I added it to stress a more generalized use case for Traverse. I hope it's still derivable from the other explanations.

### To fold Or not to fold

Even though, `Foldable` and `Traverse` are related, because both 'reduce their values to something', it is not obvious why to consider `Traverse` over `Foldable`.

Here is a small example:

```kotlin:ank:playground
import arrow.core.MapK
import arrow.core.Option
import arrow.core.Some
import arrow.core.extensions.option.applicative.applicative
import arrow.core.fix
import arrow.core.k

fun main(){
  //sampleStart
  val map: MapK<String, Int> = mapOf("one" to 1, "two" to 2).k()

  val optionMapK: Option<MapK<String, String>> = map.traverse(Option.applicative()) { Some("$it") }.fix()
  

>   // val optionMapKBoilered = map.foldLeft(Some(emptyMap())) { acc: Option<MapK<String, String>>, i: Int ->

Author: What do you mean exactly, I didn't want to implement the fold version, because the result is just a nasty way to ducktape a new `MapK`

  //   acc.fold({ emptyMap() }, { /*Some logic to retrieve the key of a value, transform it and add it to the accumulated Map*/ })
  // }
  //sampleEnd
  println(optionMapK)
}
```

> Both values try to attain the same thing, but what `Foldable` lacks is that it solely drills down to it's `A` here `Int` and does not preserve it's shape `F` - `MapK`. This is where `Traverse` shines, whenever you care about the Output `B` from `(A) -> B` and the existing shape of `F` you would use `traverse`.

Reviewer 5: ```suggestion
Both values try to attain the same thing, but what `Foldable` lacks is that it solely drills down to it's `A` here `Int` and does not preserve it shapes `F` - `MapK`. This is where `Traverse` shines, whenever you care about the Output `B` from `(A) -> B` and the existing shape of `F` you would use `traverse`.
```

Author: ```suggestion
Both values try to attain the same thing, but what `Foldable` lacks is that it solely drills down to it's `A` here `Int` and does not preserve its shape `F` - `MapK`. This is where `Traverse` shines, whenever you care about the Output `B` from `(A) -> B` and the existing shape of `F` you would use `traverse`.
```
Thx, I meant `its` instead of `it's`

Additionally, you're able to wrap your context `F` within a `G`. That is, may among others, the reason why `Traverse` is strictly more powerful than `Foldable`.

> You can also misuse it's powers - where a `map` is more considerable: 

Reviewer 2: What do you mean with this?

Author: Here I wanted to give a counter-example, how not to use `Traverse`. Here it results in an inefficient implementation because we `wrap` and `unwrap` the `Int`

Author: I think I had to add that exact sentence from above to the markdown, haha

  val isLeft = listOf("1", "abc", "3").k().traverse(Either.applicative(), ::parseIntEither).isEmpty()
  //sampleEnd
  println("list= $list")
  println("isLeft= $isLeft")
}
```

```kotlin:ank:playground
import arrow.core.Either
import arrow.core.Nel
import arrow.core.ValidatedNel
import arrow.core.extensions.nonemptylist.semigroup.semigroup
import arrow.core.extensions.validated.applicative.applicative
import arrow.core.fix
import arrow.core.invalidNel
import arrow.core.k
import arrow.core.validNel

fun parseIntEither(s: String): Either<NumberFormatException, Int> =
  if (s.matches(Regex("-?[0-9]+"))) Either.Right(s.toInt())
  else Either.Left(NumberFormatException("$s is not a valid integer."))

fun parseIntValidated(s: String): ValidatedNel<NumberFormatException, Int> =
  if (s.matches(Regex("-?[0-9]+"))) s.toInt().validNel()
  else NumberFormatException("$s is not a valid integer.").invalidNel()

fun main() {
  //sampleStart
  val validatedList = listOf("1", "22w", "3", "33s").k()
    .traverse(ValidatedNel.applicative(Nel.semigroup<NumberFormatException>()), ::parseIntValidated).fix()
  //sampleEnd
  println(validatedList)
}
```

> Notice that in the `Either` case, should any string fail to parse the entire `traverse` is considered a failure. Moreover, once it hits its first bad parse, it will not attempt to parse any others down the line (similar behavior would be found with using `Option`). Contrast this with `Validated` where even if one bad parse is hit, it will continue trying to parse the others, accumulating any and all errors as it goes. The behavior of `traverse` is closely tied with the `Applicative` behavior of the data type.

Reviewer 2: I like these examples 👍
